### ISSUE TITLE
Config Server+Vault: Vault token renewing fails

### ISSUE THREAD

**Issue by RushPVA** [pivotal-cf/spring-cloud-services-connector#44]
2018-04-25T21:09:53.853+03:00 [APP/PROC/WEB/0] [OUT] 2018-04-25 18:09:53.852 INFO 14 --- [pool-7-thread-1] c.s.c.VaultTokenRenewalAutoConfiguration : Renewing Vault token e0d9[*]57cf for 300000 milliseconds.
2018-04-25T21:09:53.898+03:00 [APP/PROC/WEB/0] [OUT] 2018-04-25 18:09:53.897 ERROR 14 --- [pool-7-thread-1] c.s.c.VaultTokenRenewalAutoConfiguration : Unable to renew Vault token e0d9[*]57cf. Is the token invalid or expired?

while debugging i found that Vault respond:
Content type 'application/xml;charset=UTF-8' not supported","path":"/vault/v1/auth/token/renew-self"

I was able to fix it by adding to io.pivotal.spring.cloud.service.config.VaultTokenRenewalAutoConfiguration#buildTokenRenewRequest

headers.setContentType(MediaType.APPLICATION_JSON);

**Comment by csterwa** [comment]
@RushPVA could you provide the version of Pivotal Cloud Foundry, Spring Cloud Services (SCS) and SCS Connector that you are using? Also, was the instance created recently or has it existed for some time?

**Comment by csterwa** [comment]
Thanks @RushPVA - I'm attempting to recreate this in PWS. I will let you know what I come up with or if we have any other questions.

**Comment by RushPVA** [comment]
@csterwa ok thanks
(may be it depends also on vault version , we use one of the latest versions )

**Comment by csterwa** [comment]
@RushPVA I was able to reproduce on PWS with SCS Connector `v1.6.3`. I was using Vault `v0.7.2. We'll take a look at why this is the case. Thanks for informing us about the issue.

**Comment by csterwa** [comment]
@RushPVA our acceptance tests pass using `v0.6.5`. We're going to work on an upgrade process for our Vault server in the near future so we can catch any oddities that arise from such an upgrade in the future.

**Comment by habuma** [comment]
@RushPVA May I ask which version of Vault you are using?

**Comment by habuma** [comment]
FWIW, I have a theory on what's happening, but without seeing the actual application code that this is happening in, it's only a theory...

Normally (e.g., in a typical Spring Boot app without any peculiar dependencies or configuration), `RestTemplate#postForObject()` will send data in JSON with a `Content-type` header of "application/json". On the receiving end, in the Config Server there is a `VaultRelayController` that accepts that request and forwards it on (with a tiny bit of tweaking) to the Vault server. (The relay controller exists so that the individual applications don't need to be aware of Vault and only need to be aware of the Config Server, which they're already aware of.) In short, because `RestTemplate` sends JSON and because the relay controller expects JSON, it "just works".

But it is possible that there is some dependency and/or configuration in the client application that causes `RestTemplate` to send XML. In fact, it is very clear from the error shown that this is definitely the issue--that XML is being sent when JSON is expected. The reason why is uncertain, but it is very clear that XML is being sent. The proposed change of setting the `Content-type` header overcomes that by forcing it to be JSON.

Another thing worth noting is that this is not likely to have anything to do with a specific version of Vault (although I did learn along the way that Vault 0.10.0 including some potentially breaking changes in the API). The request to "/vault/v1/auth/token/renew-self" in the reported error is for the token relay controller on the config server, not an endpoint on Vault API. 

Put simply, the relay server is not equipped to handle XML in the POST request, but (for some reason) the client is sending the renewal request as XML instead of JSON. 

The proposed solution of explicitly setting the `Content-type` is a good one, but I'd like to understand *why* XML is being sent before making such a change. Is it possible that there is something unusual in the classpath that puts `RestTemplate` into a mode of sending XML by default? Or some explicit configuration perhaps?

**Comment by RushPVA** [comment]
@habuma , vault version is 0.8.3

**Comment by RushPVA** [comment]
about RestTemplate:
default ContentType depends on list of converters(we use default list that is created in default constructor of RestTempalte) and depends on type of request(in our case it is HashMap, you can check io.pivotal.spring.cloud.service.config.VaultTokenRenewalAutoConfiguration#buildTokenRenewRequest)

default list of converters
```
	public RestTemplate() {
		this.messageConverters.add(new ByteArrayHttpMessageConverter());
		this.messageConverters.add(new StringHttpMessageConverter());
		this.messageConverters.add(new ResourceHttpMessageConverter());
		this.messageConverters.add(new SourceHttpMessageConverter<Source>());
		this.messageConverters.add(new AllEncompassingFormHttpMessageConverter());

		if (romePresent) {
			this.messageConverters.add(new AtomFeedHttpMessageConverter());
			this.messageConverters.add(new RssChannelHttpMessageConverter());
		}

		if (jackson2XmlPresent) {
			this.messageConverters.add(new MappingJackson2XmlHttpMessageConverter());
		}
		else if (jaxb2Present) {
			this.messageConverters.add(new Jaxb2RootElementHttpMessageConverter());
		}

		if (jackson2Present) {
			this.messageConverters.add(new MappingJackson2HttpMessageConverter());
		}
		else if (gsonPresent) {
			this.messageConverters.add(new GsonHttpMessageConverter());
		}
	}
```

the first converter from this list that can marshall our request will define ContentType
in our case it is MappingJackson2XmlHttpMessageConverter

so this converter set ContentType header in  'application/xml;charset=UTF-8'

**Comment by habuma** [comment]
Gotcha. I understand how message converters work, which is why I wanted to know if there was something in your classpath that might cause one of the XML-based message converters to take precedence. I figured it was something like that, but didn't want to move forward on a guess. Now that it's not a guess, I can make that fix.

**Comment by habuma** [comment]
Per the original issue in tracker (https://github.com/pivotal-cf/spring-cloud-services-connector/issues/44) this has been delivered, but not accepted. This may have gotten lost in the shuffle as we transitioned from Tracker to Waffle. Shall I close this?
